### DISCUSSION TITLE
get the content out of Buffer.t

### DISCUSSION THREAD

**Discussion by c-cube** [ocaml/ocaml#12881] [Features]
I'm still annoyed at the interface of `Buffer.t` because it forces me to copy the entire content to a string in order to access it. In the past I proposed to make the content public, which was rejected because it'd prevent future changes to the internal representation.

Would maintainers accept a less ambitious PR that adds this?

```ocaml
(* buffer.mli *)

(** Iterate on chunk(s) of data inside the buffer, in the order in which the data was
     added. The number of chunks is not specified. *)
val iter_chunks : (bytes -> int -> int -> unit) -> t -> unit
```

This subsumes things like `Buffer.output_buffer` which are sadly not sufficient because output channels are not extensible.

**Comment by damiendoligez** [comment]
FWIW, I think this is a good idea.

**Comment by gasche** [comment]
I agree that this approach is robust to changes of representations, so I'm fine with the proposed interface.

**Reply by alainfrisch** [reply]
The proposal would support only some changes of internal representation, but not, for instance, using bigarrays (or more "native" kinds of buffers for JavaScript backends). I'm not opposed to the change at all, it's just that I don't find the argument about possible change of representation very strong.
